### PR TITLE
Design Document for Reading Compressed Objects from S3

### PR DESCRIPTION
**Pull request by cirego**
After the discussion on #6048, it seemed like it's worth trying the new design doc review process for this feature. We also have confirmation, via the customer and reading Kafka's S3 Connect source code, that we do need to add a `COMPRESSION {}` specification to S3 sources. The `Content-Encoding` header is not set reliably.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6057)

### CODE REVIEW

**Comment by benesch** [comment]
Everything that’s in here seems great, but would love to see a few sentences about what happens if content encoding is set and does not match an explicit compression specification from the user.

**Comment by cirego** [comment]
> Everything that’s in here seems great, but would love to see a few sentences about what happens if content encoding is set and does not match an explicit compression specification from the user.

Thanks! I updated the document to say that mismatches will not affect normal operation but will be logged at the debug level.

File: doc/developer/design/20210310_compressed_s3_sources.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+Materialize will attempt to decompress the object using gzip, and if that fails, will attempt to
+treat the file as uncompressed. Files that cannot be parsed, even after attempting decompression,
+will result in errors in the error stream for the source.
+
+## Additional Testing
+
+Testing will be done via testdrive. See #6048 for an example on how to test object compression
+with S3.
+
+## Alternatives
```
Comment (quodlibetor) [inline doc/developer/design/20210310_compressed_s3_sources.md:71]: There is a fourth alternative -- which I'm definitely not advocating for -- of examining the object downloaded for gzip/deflate/brotli headers and using that to do the decompression. The biggest reason to not do that to that is that we're not sure how many false positives examining headers would create.

**Review by quodlibetor — APPROVED** [review]

**Comment by benesch** [comment]
@quodlibetor, what's the status of the "object manifest" we were talking about originally for S3 sources? One could imagine that one of the fields of the manifest specified what compression algorithm was in use for each object. Not sure how common using multiple compression algorithms in the same source would be, though. Mostly just food for thought.

**Comment by cirego** [comment]
> @quodlibetor, what's the status of the "object manifest" we were talking about originally for S3 sources? One could imagine that one of the fields of the manifest specified what compression algorithm was in use for each object. Not sure how common using multiple compression algorithms in the same source would be, though. Mostly just food for thought.

Is object manifest something that I should add to the list of alternatives?

File: doc/developer/design/20210310_compressed_s3_sources.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+`COMPRESSION` is specified for files.
+
+### Compression "NONE"
+
+Materialize will not try to decompress any objects downloaded from this S3 source. Files that
+cannot be parsed, either due to being compressed or for another reason, will result in errors in
+the error stream for the source.
+
+The `Content-Type` field will be ignored for functional purposes but verified. If the
+`Content-Type` is not `identity`, a debug message will be generated indicating the mismatch.
```
[the remark this answers is not in the document]
Comment (cirego) [reply doc/developer/design/20210310_compressed_s3_sources.md:37]: Yes, good catch. Thank you!

File: doc/developer/design/20210310_compressed_s3_sources.md (hunk 1 above)
Comment (cirego) [reply doc/developer/design/20210310_compressed_s3_sources.md:71]: I'll go ahead and add it!

**Comment by cirego** [comment]
There are two main questions that I would appreciate help answering:

1. Do we want to implement `AUTO` or should we push that until a customer asks for it? My current leaning is that we should wait until a customer tries the feature and requests AUTO before implementing it. Partly because that will also help inform question 2:

2. How do we handle mismatched between `Content-Encoding` and the compression setting? My current leaning is that we should ignore `Content-Encoding`, as we know that there are examples of producers that will generate compressed data without setting this header.

**Comment by elindsey** [comment]
> There are two main questions that I would appreciate help answering:
> 
> 1. Do we want to implement `AUTO` or should we push that until a customer asks for it? My current leaning is that we should wait until a customer tries the feature and requests AUTO before implementing it. Partly because that will also help inform question 2:
> 2. How do we handle mismatched between `Content-Encoding` and the compression setting? My current leaning is that we should ignore `Content-Encoding`, as we know that there are examples of producers that will generate compressed data without setting this header.

For what it's worth, I 100% agree on how you're leaning on both points. AUTO always gets messy (is content encoding a lie? is content encoding set at all? did they accidentally double gzip?), and I'd rather start with something concrete and minimal like explicit gzip opt-in.

**Comment by quodlibetor** [comment]
> what's the status of the "object manifest" we were talking about originally for S3 sources?

We decided to punt on it for awhile with the general feeling being that we didn't want to implement sources that are that dynamic/complex until we had a strong customer need.

> Is object manifest something that I should add to the list of alternatives?

It's probably worth calling out and linking to the issue (https://github.com/MaterializeInc/materialize/issues/5502), yeah.

> My current leaning is that we should ignore Content-Encoding

yeah, the conversation here has convinced me that we should ignore it (possibly with a debug-level log if it doesn't match).

**Comment by benesch** [comment]
👍🏽 very on board with only using explicit compression specs for now.

**Comment by cirego** [comment]
Sounds like we're all in agreement on the explicit compression handling behavior and (not) supporting Content Encoding. Thanks everyone!

**Pull request merged**